runtime/src/iree/base/internal/cpu.c

```diff
@@ -32,38 +32,133 @@
+                 IREE_HWCAP2_I8MM);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_BF16, hwcap2,
+                 IREE_HWCAP2_BF16);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SVE, hwcap,
+                 IREE_HWCAP_SVE);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SVE2, hwcap2,
+                 IREE_HWCAP2_SVE2);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SVE2P1, hwcap2,
+                 IREE_HWCAP2_SVE2P1);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SVE2_BITPERM, hwcap2,
+                 IREE_HWCAP2_SVEBITPERM);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_F32MM, hwcap2,
+                 IREE_HWCAP2_SVEF32MM);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_F64MM, hwcap2,
+                 IREE_HWCAP2_SVEF64MM);
+
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME, hwcap2,
+                 IREE_HWCAP2_SME);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME2, hwcap2,
+                 IREE_HWCAP2_SME2);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME2P1, hwcap2,
+                 IREE_HWCAP2_SME2P1);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME_F16F16, hwcap2,
+                 IREE_HWCAP2_SME_F16F16);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME_F64F64, hwcap2,
+                 IREE_HWCAP2_SME_F64F64);
+  IREE_COPY_BITS(out_fields[0], IREE_CPU_DATA0_ARM_64_SME_I16I64, hwcap2,
+                 IREE_HWCAP2_SME_I16I64);
 }
 
 #elif defined(IREE_PLATFORM_MACOS) || defined(IREE_PLATFORM_IOS)
 
 #include <sys/sysctl.h>
 #include <sys/types.h>
 
-#define IREE_QUERY_SYSCTL(key, field_value, field_bit)            \
-  do {                                                            \
-    int64_t result = 0;                                           \
-    size_t result_size = sizeof result;                           \
-    if (0 == sysctlbyname(key, &result, &result_size, NULL, 0)) { \
-      if (result) field_value |= field_bit;                       \
-    }                                                             \
-  } while (0)
-
 static void iree_cpu_initialize_from_platform_arm_64(uint64_t* out_fields) {
-  IREE_QUERY_SYSCTL("hw.optional.arm.FEAT_DotProd", out_fields[0],
-                    IREE_CPU_DATA0_ARM_64_DOTPROD);
-  IREE_QUERY_SYSCTL("hw.optional.arm.FEAT_I8MM", out_fields[0],
-                    IREE_CPU_DATA0_ARM_64_I8MM);
+  typedef struct {
+    const char* sysctl_key;
+    int out_field_index;
+    uint64_t out_field_bits;
+  } feature_t;
+  const feature_t features[] = {
+      {
+          .sysctl_key = "hw.optional.floatingpoint",
+          .out_field_index = 0,
+          .out_field_bits = IREE_CPU_DATA0_ARM_64_FP_ARMV8,
+      },
+      {
+          .sysctl_key = "hw.optional.arm.FEAT_LSE",
+          .out_field_index = 0,
+          .out_field_bits = IREE_CPU_DATA0_ARM_64_LSE,
+      },
+      {
+          .sysctl_key = "hw.optional.arm.FEAT_LSE2",
```

Comment:
Should this be 128? (I'm not sure how close we want to keep the key to the macro)

Reply:
Good question! Actually, there is discrepancy in how different platforms call the same Arm architecture features. `FEAT_LSE2` is how Apple calls it, and also how Arm itself calls it. "lse128" is how LLVM calls it. So this is intentional.